### PR TITLE
[Train] Split test_gpu tests

### LINKED ISSUE #29299: [CI] `linux://python/ray/train:test_gpu` is failing/flaky on release branch.
- 4b605c24aeb991e4b693f8fa37fc967079d3ec22 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/498#0183ce28-68b0-4e81-b4d4-63909e9df9d7)
- 6bac4ab766d3c738bb816cb5fd6801aaab4388f4 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/495#0183cded-4ab5-4dd6-b3ea-d488a6604c53)
- 429c0e88cd0bc96f75e5125525bd660232eddcbb FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/493#0183cda3-d271-4b0d-ade0-be1a3a2229d8)
- 40811d7029b8ce721713a3532596e1aba9646ab0 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/492#0183cd9f-f951-4ee9-aa2a-a889e27622f7)
- 34cd1cbccf295db77fcfb5f81421e295f2664643 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/487#0183cd52-77e7-4140-a1b5-c13783900adb)
- f2dacc017c066f975ef24c67eed0411cdfac12d2 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/486#0183cd4f-7d35-42b6-aec5-ed91d9c870d9)
- 223ce3988e0a424d33a557c616eff6dd140b812c FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/482#0183cb57-20a6-4b27-b149-7bb983003ecc)
- 75e9722a4d9c0d8d5cb0c37eb6316553f9e0789e FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/481#0183cac0-74c0-4a96-81aa-3d12a9f1f775)
- 49a27f2b52475142a08fa771637b3b5a01fc1701 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/480#0183ca4f-07cb-476c-ae42-7716f96d0e94)
- 5229df15ab71baf22ade6a3ff14e921be10b3a25 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/477#0183c9b8-4f8a-4da9-a400-d545e4e05ca5)
- 44fc6a36e5228a50f4bafe126c67ad73d78486eb FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/474#0183c991-ac6c-4902-892e-91c3e0c034b3)
- 7d3eeb13032f341627015353214524fe09fad6a6 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/473#0183c98c-2485-4009-920d-268113d2d82c)
- 4b79a6f1d270658eca3bf65df2b16b4e512f6513 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/469#0183c94c-65f4-4a09-81aa-7b5d71144f3f)
- c9875edafc167d70c55fef54cc045f14f39b19a5 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/468#0183c946-b088-4339-8b31-9ff075d23f37)
- 60840b5498a8e5e8fc6b1bbeadaaa5c210b62624 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/465#0183c901-1478-42a4-a5ed-4b80ddb78d1e)
- 297341e107daee1ea3aff991ae8ea8c90993c683 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/460#0183c868-9f50-4989-b153-f8e9bf4debeb)
- d89a66401719996b3f56b2769ca24d5fc20aa889 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/455#0183c831-7593-4a19-a91b-9e120d4846fe)
- 217dcd7f18cd4ce0025a68bd46f0bf249c09870d FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/454#0183c806-61cd-4ac6-8344-66cbb2de223a)
- 217dcd7f18cd4ce0025a68bd46f0bf249c09870d FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/451#0183c4e2-7244-427e-b931-3e3e68530d74)
- 1bd8ab9b421a58cd07fe4ca015c07e5fd0070059 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/450#0183c4b1-49f5-4d45-911b-4a6e91fd5f85)
- 91040d91bba3cdfc29475b77e8d199884469ab86 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/379#0183aefd-576f-4d74-b455-3eced0ab5e47)
- 8486ddd1bf909449a22262c52a90f8359797fc16 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/378#0183aef8-ee16-4cf5-9fad-c326f0b5e8d0)
- 7e939d79bcc68603b7c9a99fa6ef880dd526e931 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/377#0183aeef-f145-4257-aa9f-ae59634ccec9)
- 4732753327109d5c33b4cf69d375730057efc631 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/376#0183aee1-2dde-48a6-8e1d-f8ffce438752)
- 68a35fceef6a47fa3e3da41041ee32dafe7f24a8 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/375#0183aec9-28c0-465b-9f05-db096f629ae1)
- 23b3a599b9df8100558c477e94b0b19b1a38ac27 FAILED [Buildkite :tv: :steam_locomotive: Train GPU tests ](https://buildkite.com/ray-project/oss-ci-build-branch/builds/374#0183aec0-8b6d-428e-a1fc-90eedbf969c2)

....
Generated from flaky test tracker. Please do not edit the signature in this section.
DataCaseName-linux://python/ray/train:test_gpu-END
....

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

We had AMI changes that happened around same time `test_gpu` times out but for our ongoing work we should keep updated with latest pytorch and cuda 11.6.

Therefore this change split tests into two files so single buck target won't timeout.

## Related issue number

Closes #29299

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by amogkam** [comment]
Can we do a split like this:

- Examples in its own files (test_torch_fashion_mnist, test_horovod, etc.)
- Auto gpu to cpu in its own file
- Everything else keep in `test_gpu` for now

**Review by krfricke — APPROVED** [review]
LGTM waiting for CI

**Comment by amogkam** [comment]
I'm all for splitting the tests, but would like us to do it in a more organized fashion. What do you think about my suggestion above?

I want to avoid getting into a situation where we have things like `test_trial_runner`, `test_trial_runner_2`, `test_trial_runner_3`, etc.

**Comment by krfricke** [comment]
Yeah, semantic splitting sounds good to me

**Review by amogkam — CHANGES_REQUESTED** [review]
Requesting changes for the comment above

**Comment by jiaodong** [comment]
> * Examples in its own files (test_torch_fashion_mnist, test_horovod, etc.)
> * Auto gpu to cpu tests in its own file
> * Everything else keep in `test_gpu` for now

last two points make sense to me, `test_gpu_examples` seems structured fine to me tho as they're all doing `mnist` gpu training from different frameworks ?

File: python/ray/train/BUILD [hunk 1]
```diff
@@ -190,6 +190,14 @@ py_test(
...
+    name = "test_gpu_with_amp",
+    size = "large",
+    srcs = ["tests/test_gpu_with_amp.py"],
+    tags = ["team:ml", "exclusive", "gpu_only"],
+    deps = [":train_lib"]
+)
+
 py_test(
     name = "test_torch_fsdp",
     size = "large",
```
Comment (amogkam) [inline python/ray/train/BUILD:203]: need to add test_gpu_auto_transfer here too

File: python/ray/train/BUILD [hunk 2]
```diff
@@ -190,6 +190,14 @@ py_test(
     deps = [":train_lib"]
 )
 
+py_test(
+    name = "test_gpu_with_amp",
```
Comment (amogkam) [inline python/ray/train/BUILD:194]: Nit: test_amp_gpu

File: python/ray/train/BUILD (hunk 1 above)
Comment (jiaodong) [reply python/ray/train/BUILD:203]: it's already there with `gpu_only` tag above

**Review by amogkam — APPROVED** [review]
Nice, thanks!

File: python/ray/train/BUILD (hunk 2 above)
Comment (amogkam) [reply python/ray/train/BUILD]: good point, sounds good to leave as is

File: python/ray/train/BUILD (hunk 1 above)
Comment (amogkam) [reply python/ray/train/BUILD:203]: ah sorry didnt see that!

**Pull request merged**